### DISCUSSION TITLE
인터프리터 언어와 컴파일 언어에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by BaekYuri** [wanted-9th-3team/frontend-study#25] [프론트 기술 질문 공간]
인터프리터 언어와 컴파일 언어에 대해 설명해주세요.

**Comment by skyhanull** [comment]
컴파일 언어는 고급언어를 저급언어로 번역하는 프로그램을 말합니다
컴파일 언어에서는 컴파일과정과 실행과정으로 나뉘는데 컴파일과정에서 먼저 컴파일러를 통해 컴파일 타임에 전체 소스 코드를 한 번에 기계어로 변환 후 실행파일을 만든다. 컴파일은 한번만 실행되고 실행과정에서는 실행만 하면 되므로 속도가 빠릅니다

인터프리터 언어는 컴파일과정이 없고 프로그래밍 언어의 소스 코드를 바로 실행하는 컴퓨터 프로그램을 말한다.
컴파일 과정이 없어 컴파일하는 시간은 사라지나 한줄한줄 읽기 때문에 실행 시 마다 인터프리터과정이 실행되어 속도가 느립니다

**Comment by michoball** [comment]
컴파일 언어는 작성된 코드를 한번에 기계어로 변환하는 컴파일 단계를 거치고 실행이 됩니다. 컴파일이 되고나면 build된  파일만을 가지고 실행이 되기 때문에 실행되는 속도는 빠르나 많은 양의 코드를 가지고 있다면 컴파일 단계에 시간이 많이 소요됩니다. 
그리고 코드를 중간에 바꾸면 전체를 다시 컴파일해야하는 단점이 있습니다.  대표 언어로는 C, C#, Java 등이 있습니다. 

인터프리터 언어는 작성된 코드를 한 줄 한 줄 가져와 읽으면서 실행을 합니다. 따라서 컴파일 단계가 없고 중간에 코드를 수정하면 그부분만 반영되어 실행이 가능하지만 전체적인 실행시간이 컴파일 언어에 비해 많이 소요됩니다.  대표 언어로는 javascript, python이 있습니다.

**Comment by WooGie911** [comment]
컴파일 언어는 컴파일러를 통해 컴파일 타임에 전체 소스 코드를 한 번에 기계어로 변환 후 
실행파일을 만든후 실행합니다. 

컴파일 언어는 컴파일 단계와 실행 단계가 각각 분리되어 있으며, 컴파일은 단 한번만 수행합니다.

실행은 실행 파일을 실행시킴으로써 할 수 있으며, 
실행시에는 컴파일 과정을 거치지 않고 실행만 하면 되므로 코드 실행 속도가 빠릅니다. 

단, 프로젝트의 규모가 클 경우 컴파일 시간이 오래 걸릴 수 있다는 단점이 있습니다.

대표적인 컴파일 언어로는 C, C++이 있습니다.


인터프리터 언어는 소스 코드를 한번에 기계어로 변환하는 컴파일러와 달리, 
컴파일 하지 않고 소스 코드를 한 줄씩 읽어들여 실행합니다. 

컴파일 하는 과정이 없기 때문에 컴파일 하는 시간은 소요되지 않으나, 
인터프리터 언어는 실행파일을 별도로 생성하지 않기 때문에 
실행시마다 인터프리트 과정이 반복 수행되어 실행 속도가 느리다는 단점이 있습니다.

대표적인 인터프리터 언어로는 R, Python, Ruby와 같은 언어들이 있습니다.

**Comment by javascriptbiggosoo** [comment]
프론트엔드 개발자들의 친구인 JavaScript의 경우 인터프리터 언어이기 때문에 여러분들이 설명해주신 것처럼 느리다는 단점이 '있었는데요 없었습니다.'
크롬 v8 엔진이 인터프리터를 던져버리고 Just In Time 컴파일러를 도입하면서 기계어로 한 번 변환시켜 준 후 코드를 실행시키기 때문에 인터프리터 언어로만 작동하던 시절보다 훨씬 빨라졌습니다.

**Comment by BaekYuri** [comment]
컴파일 언어 : 작성한 프로그래밍 언어를 기계어로 컴파일하고 이를 실행합니다. 컴파일 하는 과정은 오래 걸리지만 컴파일 후 실행 속도는 매우 빠른 편입니다. OS 이식성이 낮기 때문에 다른 OS로 바뀐다면 다시 컴파일 해야합니다.
인터프리터 언어 : 소스 코드를 인터프리터를 통해 한 줄씩 번역 및 실행을 동시에 실행합니다. 줄 단위로 번역 및 실행을 하기 때문에 속도는 조금 느린 편이지만 OS마다 호환되는 인터프리가 있다면 실행할 수 있으므로 OS 이식성이 높은 편입니다. 또한 어떤 코드에서 오류가 생긴다면, 그 아래의 코드들은 실행되지 않기 때문에 디버깅에 용이한 편입니다.

**Comment by bona373737** [comment]
인터프리터 언어 : 코드가 실행 단계인 런타임에 코드 한 줄씩 중간 코드인 바이트코드로 변환 후 실행.
컴파일 하는 과정이 없기 때문에 컴파일에 소요되는 시간은 없지만 인터프리터 언어는 실행파일을 별도로 생성하지 않기 때문에 실행시마다 인터프리트 과정이 반복 수행되어 실행 속도가 느리다는 단점이 있습니다.

컴파일언어 : 코드 실행 전 컴파일 타임에 소스 코드 전체를 한번에 기계어로 변환 후 실행.
컴파일과 실행단계가 분리되어 있어, 실행시에는 실행만 하면 되므로 코드 실행 속도가 빠릅니다.
